cctp_primitives/src/proving_system/init.rs

```diff
@@ -18,104 +13,204 @@ use std::sync::{RwLock, RwLockReadGuard};
+    pub static ref G2_UNIVERSAL_PARAMS: RwLock<Option<UniversalParams<G2>>> = RwLock::new(None);
 }
 
-/// Generate G1CommitterKey and store it in memory.
+/// Generate `G1_UNIVERSAL_PARAMETERS` and store it in memory.
+/// This function should be called exactly once during program execution and before any call to
+/// `get_g1_committer_key()`. Further calls leave `G1_UNIVERSAL_PARAMETERS` unchanged and return an
+/// error instead.
 /// The parameter `max_degree` is required in order to derive a unique hash for the key itself.
-pub fn load_g1_committer_key(
-    max_degree: usize,
-    supported_degree: usize,
-) -> Result<(), SerializationError> {
-    match load_generators::<G1>(max_degree, supported_degree) {
+pub fn load_g1_universal_params(max_degree: usize) -> Result<(), SerializationError> {
+    {
+        let pp_g1_guard = G1_UNIVERSAL_PARAMS.read().map_err(|_| {
+            SerializationError::IoError(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "Failed to acquire lock for G1_UNIVERSAL_PARAMS".to_owned(),
+            ))
+        })?;
+        if pp_g1_guard.is_some() {
+            return Err(SerializationError::IoError(std::io::Error::new(
+                std::io::ErrorKind::AlreadyExists,
+                "G1_UNIVERSAL_PARAMS has already been generated",
+            )));
+        }
+    }
+    match load_universal_params::<G1>(max_degree) {
         // Generation/Loading successfull, assign the key to the lazy_static
-        Ok(loaded_key) => {
-            G1_COMMITTER_KEY
+        Ok(loaded_params) => {
+            G1_UNIVERSAL_PARAMS
                 .write()
                 .as_mut()
                 .map_err(|_| {
                     SerializationError::IoError(std::io::Error::new(
                         std::io::ErrorKind::Other,
-                        "G1_COMMITTER_KEY write failed",
+                        "G1_UNIVERSAL_PARAMS write failed",
                     ))
                 })?
-                .replace(loaded_key);
+                .replace(loaded_params);
             Ok(())
         }
         // Error while generating/reading file/writing file
         Err(e) => Err(e),
     }
 }
 
-/// Generate G2CommitterKey and store it in memory.
+/// Generate `G1_UNIVERSAL_PARAMETERS` and store it in memory.
+/// This function should be called exactly once during program execution and before any call to
+/// `get_g1_committer_key()`. Further calls leave `G1_UNIVERSAL_PARAMETERS` unchanged and return an
+/// error instead.
 /// The parameter `max_degree` is required in order to derive a unique hash for the key itself.
-pub fn load_g2_committer_key(
-    max_degree: usize,
-    supported_degree: usize,
-) -> Result<(), SerializationError> {
-    match load_generators::<G2>(max_degree, supported_degree) {
+pub fn load_g2_universal_params(max_degree: usize) -> Result<(), SerializationError> {
```

Comment:
Please let's keep the old name anyway to be sure not to break dependant libraries (formally, doesn't change anything anyway)